### ISSUE TITLE
Localisation: General info

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#148]
This is a thread for the **general** discussion of localising OptiKey, e.g. the discussion of how the process works, which languages should be done now/next and branches should be created, etc. It is not the place to discuss specific languages - we can create new issues for those - just ask here.

A full guide has been written which explains the technical process of localising OptiKey into a new language. It is available [here](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md).

A less technical, more "overview" guide is as follows:

**1.Translation of the UI strings** (the strings used on keys in the user interface, e.g. "MENU" and "SPEAK") need to be translated on the Transifex site: https://www.transifex.com/optikey/ - you need to register (for free) on that site and then request to join as a translator for each language that you want to help with. I need to approve your applications and then you're good to go. Please don't request English as a new language on Transifex - it's not necessary.

**2.We need information about what the keyboard looks like.** This website has great info on the most common layouts: http://ascii-table.com/keyboards.php
Screenshots of keyboard layouts like these are also very helpful: http://kbd-intl.narod.ru/english/layouts

**3.Alphabet** A list of upper and lower letters is required - these are stand alone letters that count as part of the core alphabet of the language. 

A list of diacritic marks which can be combined with letters is also required. For each mark (e.g. an acute accent) a list of associated letters should be supplied. This may not make sense to you, but some languages, such as Turkish, treat some letters with diacritic marks as stand alone letters, whilst other letters can be decorated with a diacritic mark, but these are not counted as stand alone letters. This site has info on most diacritics by language: http://diacritics.typo.cz/index.php?id=49

Are there any other special characters?
N.B. Ligatures (e.g. œ and æ) are NOT supported - they are a bloody nightmare!

**4.A word list** - this is the dictionary that OptiKey uses for auto-complete and multi-key selection. It should meet the following critieria:

i)It should be roughly 75k-100k words in size (it can be a bit less or more, but there is a performance overhead for huge dictionaries so there is no point choosing one that is 1 million words long).
ii)No entries of 1 character as the user can type these from the keyboard and they will be ignored.
iii)All words should be lower case, unless the case has actual meaning, e.g. proper nouns like names, abbreviations like "ATM", or some other meaningful use of capital letters. OptiKey will case words automatically otherwise.
iv)There is no need to manually inspect the whole word list, but sanity check it for mistakes, swear words and colloquialisms that wouldn't be used by the average person. It also doesn't need to contain obscure words. Generally speaking it should cover a normal vocabulary for an adult speaker.
v)If your dictionary includes ligatures (e.g. "œ" and "æ") then please convert them into the individual letters (e.g. "oe" and "ae"). If you think this is not appropriate please contact me to discuss options.
vi)The dictionary files have to be encoded in unicode (UTF-8). If you encode them in ASCII some characters (e.g. ü) will instead be displayed as the 'replacement character' (�). To do this just copy/paste the whole word list into Notepad and save the file with the Encoding set to "UTF-8" like this;
![image](https://cloud.githubusercontent.com/assets/4722561/10235169/06dfce82-6892-11e5-9703-fdbf1b8549db.png)

Some word lists are available here, but please check the length as they are massive: http://www.winedt.org/dict.html

This site also has word lists built from the most common words found in tv and film scripts: https://invokeit.wordpress.com/frequency-word-lists/ - I have written a script which can extract the top X from these lists if they are good quality for a specific language, but please note that these lists tend to include a lot of duplicates (with different accents/other diacritics) and have generally been pretty poor quality.

**5.Anything else?** Some languages have other considerations that we need to think about. Is it written right to left? Does it use characters from a non-latin alphabet? etc...

**Comment by JuliusSweetland** [comment]
**Next language?** French & German have been released so we need to look to the next language. Currently Russian is in progress (on the 'Russian' branch), but I think we have capacity to also develop one more language. Here are the complete languages on Transifex:

Dutch (nl)
Portuguese (pt)
Spanish (es)
Italian (it)

My thinking would be to go for Spanish as I can't think of any potential problems and it will probably lead quite naturally onto Portugese so we could get 2 languages out in quick succession. Open to feedback...

**Comment by lukeses** [comment]
Hi,
@JuliusSweetland thank you for a great work.

I would like to contribute by creating polish version of OptiKey.
1. Polish project on transifex goes well. We have still 200 words to translate, but I promise Polish translation on transifex will be done in a week.
2. We have some special characters.
Polish keyboard:
![pl-programmers](https://cloud.githubusercontent.com/assets/4325715/10680606/912f1e22-7921-11e5-964e-4a4787fd7dfd.png)
3. On monday I will find proper polish dictionary. I will check, if it meets your suggestions.
4. I hope that polish language don't have any special problems, but will see.

**Comment by JuliusSweetland** [comment]
@lukeses Sounds good, thanks. I can create a branch for Polish when the Transifex translation is complete. In the mean time the word list is probably the most important thing, followed by a new keyboard layout. The obvious issue with the sample keyboard layout is that there appear to be 18 special characters, which will add 4-6 columns to each row if we include them all. We need to consider if they are all required.

**Comment by JuliusSweetland** [comment]
**Issue when publishing some keys where the OS language is set to a language other than English**

When I have been experimenting with OptiKey when I have my Windows language set to other languages, e.g. French, I have noticed that the output from OptiKey into other applications is wrong for keys such as ".". It doesn't matter what language OptiKey is set to, only the OS language as this affects how Windows expects the physical keyboard to be laid out. The explanation is...

For some keys I send plain text, but for others I send virtual key codes which is akin to pressing the physical keys. The advantage of virtual key codes is that the user can perform actions that rely on the actual keys being pressed, e.g. shortcuts like CTRL+P to print. The downside is that in some language keyboard layouts, like French, the output of the key is affected by shift, which is not reflected in OptiKey.

My workaround could be to send text for keys such as ".", but I'm in two minds about what to do for the other keys. This is my thinking (the relevant method is CharExtensions.ToVirtualKeyCode()):

1.Continue publishing virtual key codes for the function keys (e.g. ALT, CTRL, SHIFT, SPACE, TAB, ENTER)

2.Continue publishing virtual key codes for letter keys as these can be useful, e.g. shortcuts. Letters don't seem to ever have an alternate value when modified with SHIFT as in latin based alphabets I think all letters have a lower and upper case.

3.Continue publishing virtual key codes for number keys as these are also useful for shortcuts. Unfortunately these DO have alternate output when SHIFT is down. At the moment OptiKey half deals with this problem by releasing auto-shift when you move to the NUMBERS & SYMBOLS keyboard (and restores it when you move back). I could just publish text for these keys and avoid the problem of outputting different symbols when SHIFT is down, e.g. on an English keyboard SHIFT+1 = "!", but I don't think this is a major problem.

4.Stop publishing virtual key codes for everything else, e.g. ".", ",", ";", etc as I would have to display the SHIFT UP/SHIFT DOWN values of these keys so that the user understands what is going on, and I think this is more trouble than it is worth. Also for French (for example) forcing the user to press SHIFT to be able to output a "." is more work, which I think is a disadvantage.

Any input would be much appreciated.

@Ar2rZ Can you comment about the Russian keyboard - how do you tell an application to print for example? In English this is CTRL+P, but the "P" key is in the same location as the "з" key on a Russian keyboard. I'm interested to find out what happens when you use OptiKey to attempt something like CTRL+P, but you have the local OS language set to Russian.

**Comment by Ar2rZ** [comment]
@JuliusSweetland i will install Russian windows on one my VM's. and will see what happening. as per reading on internet is using ctrl+p under ctrl+з. but this is on physical keyboard, which has all appropriate drivers.

**Comment by JuliusSweetland** [comment]
You should be able to install the Russian language pack on your current
installation and then switch to it for testing. Saves you the trouble of
installing a whole new version

**Comment by Ar2rZ** [comment]
you actually right. i have Russian installed on my system. so, it does not matter what language i am using even if Arabic, the short keys are still same. but in OptiKey there is no handling for those short keys cases.

**Comment by JuliusSweetland** [comment]
It's easy enough for me to map the Russian keys back in, but I'd need to
know which Russian letter is which key in English as the virtual key codes
are all based on the Latin alphabet. For example which Russian character
would you expect to press with ctrl to print a document and would you be
looking for a Russian character in that example, or would your physical
keyboard also have a P which you would look for when you wanted to print?

**Comment by Ar2rZ** [comment]
i understood your question. i will talk to users of Russian keyboard, as i never used Russian keyboard. i will give you an answer some time later.

**Comment by JuliusSweetland** [comment]
No rush, it's an interesting problem that I'd like to address at some point.

**Comment by amartel** [comment]
Hi Julius,
From my point of view, I think there are situations where we have to customize the keyboard layout to help a specific user. For example, my friend Cédric requested to have specific shortcuts like @ or / in the french alpha keyboard (to be able to enter an email address or a web site url without having to change the keyboard or to use the SHIFT key). As he suffered from ALS, his main use of optikey is to be able to browse internet, write short messages (facebook or tweeter) and other basic actions. I don't think he will use key combinations like ALT+S or CTRL+ALT+DEL so I prefer to have something very simple but easy to customize. 
About your proposals:
1 => I agree. ALT, TAB, ENTER and all function keys need to use virtual key codes
2 => It's up to you. French keyboard will work perfectly but it not required in my use case
3 => same thing.
4 => YES!! It will be a great improvement

I will test v2.2.1 as soon as possible but it seems to be pretty good for me...

Kind regards,
Arnaud

**Comment by JuliusSweetland** [comment]
@amartel Thanks for your input. FYI I intend to create a new "web browsing" keyboard (I'll create an issue in case anyone wants to help with that?).

In v2.2.0 I added the ability to switch between all keyboards with one key selection, rather than having to move via the menu every time. I've had a look at your 'cederic' branch and I think you can automatically/easily update from my master branch if you want to take a look? That might remove the need to add "@" and "/" to the main ALPHA keyboard?

**Comment by amartel** [comment]
Julius,

Well, merging my branch was very easy ;-). I’m looking right now to integrate the requests from my friend but your « web browsing keyboard » may be a good answer. In fact, I just added to the « collapsed mouse menu » the 3 following items:
- Calibrate
- sleep
- magnetic cursor (he found this option VERY useful)

I think that, now, I need some time to find the best customization for Cédric. The ability to launch calibration with a key sequence (ALT+C, for example) will be my next challenge. This is the last important thing that Cédric needs…

Kind regards,
Arnaud
——————————————————————
ACLNR: Arnaud, Cécile, Lucile, Nicolas et Rémi

Le 23 octobre 2015 à 19:14:25, Julius (notifications@github.com) a écrit:

@amartel Thanks for your input. FYI I intend to create a new "web browsing" keyboard (I'll create an issue in case anyone wants to help with that?).

In v2.2.0 I added the ability to switch between all keyboards with one key selection, rather than having to move via the menu every time. I've had a look at your 'cederic' branch and I think you can automatically/easily update from my master branch if you want to take a look? That might remove the need to add "@" and "/" to the main ALPHA keyboard?

**Comment by JuliusSweetland** [comment]
> The ability to launch calibration with a key sequence (ALT+C, for example) will be my next challenge. This is the last important thing that Cédric needs…

Re-calibrate is available on the menu, so should always be one selection away. I'm glad the merge was easy.

I've added a new issue to create the web browsing keyboard: https://github.com/JuliusSweetland/OptiKey/issues/152

Out of interest what does he use to the magnetic cursor functionality for?

**Comment by amartel** [comment]
Hi Julius,

Magnetic cursor is very useful to manage the eye tracker: When my friend moves sometime his head, the eyetribe looses precision. With magnetic cursor, Cedric knows exactly the gap between "where he looks" and "where the eyetribe thinks he looks". So he can move his head again to correct the precision or/and make a calibration before complely loosing control…
I was watching TV wednesday with him and some other friends (PSG vs REAL MADRID) and he was also using his computer during the all the time (watching the other match with Manchester United or reading Facebook/tweeter). We never had to help him but, as he moves his head very often, he used magnetic cursor to know when a calibration was required. With this feauture, he can perfectly use his computer during all the day, even when he is alone (he hasn’t an assistant full-time and his wife and his 2 daughters are not at home from 8:00 to 17:00…).

I’m thinking also to integrate optikey with SARAH (http://jpencausse.github.io/SARAH-Documentation/?lang=en) as we installed this software for him several monthes ago. It allowed Cedric to command TV, WiFi smart plugs and full other things with voice. It worked pretty good but Cédric has, now, difficulties to speak loudly and voice recognition becomes impossible :-( As there is an API in SARAH to send text (using encoded urls), I think it will be very easy to add a « Send text to SARAH »option in Optikey. This way, he can still use all the commands we configured in SARAH...

Have a nice day,
Arnaud
——————————————————————
ACLNR: Arnaud, Cécile, Lucile, Nicolas et Rémi

Le 23 octobre 2015 à 23:21:46, Julius (notifications@github.com) a écrit:

The ability to launch calibration with a key sequence (ALT+C, for example) will be my next challenge. This is the last important thing that Cédric needs…

Re-calibrate is available on the menu, so should always be one selection away. I'm glad the merge was easy.

I've added a new issue to create the web browsing keyboard: #152

Out of interest what does he use to the magnetic cursor functionality for?

**Comment by JuliusSweetland** [comment]
@amartel I've added a new issue for home automation if you'd like to discuss this further: https://github.com/JuliusSweetland/OptiKey/issues/158

This was something I have had on my TODO list for a while.

**Comment by ninehundred1** [comment]
How to sign up for this threat?

**Comment by JuliusSweetland** [comment]
You may receive notifications automatically, or you may have to subscribe
to this issue?

**Comment by JuliusSweetland** [comment]
**Russian** Pre-release #1 available here: https://github.com/JuliusSweetland/OptiKey/releases/tag/v2.2.2_Russian_pre-release_1

**Comment by felipemaciel** [comment]
I want help you translate to portuguese.

**Comment by willianfatec** [comment]
@felipemaciel Fiz boa parte da tradução das strings no transifex, você pode dar uma conferida. Se quiser trocar alguma idéia sobre Eye Tracking é só falar. 

I had translated almost all string set in transifex. You can saw there and help to fix any errors that you find. If you want talk about Eye Tracking you can contact me.

**Comment by JuliusSweetland** [comment]
Thanks @willianfatec 

Regarding general progress - I have been quite busy recently with work and also applying for a grant to be able to dedicate more time to OptiKey. Russian is in progress on the 'Russian' branch and it has made me realise that I am doing a couple of things poorly, so I'm going to focus on getting those fixed and then we can press on.

**Comment by felipemaciel** [comment]
@willianfatec it's pt_PT? Maybe I should translate in pt_BR, and make adjusts and minor changes based on yours.

**Comment by JuliusSweetland** [comment]
Is it possible to cater for both Brazil and Portugal with one translation, or are the dialects too different? I'm generally hoping we can have only one version of each language, but it might not be possible.

**Comment by willianfatec** [comment]
It's pt_BR, but feel free to make any changes. Contributions are always
welcome.

2015-11-03 15:20 GMT-02:00 felipemaciel <notifications@github.com>:

> @willianfatec <https://github.com/willianfatec> it's pt_PT? Maybe I
> should translate in pt_BR, and make adjusts and minor changes based on
> yours.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliusSweetland/OptiKey/issues/148#issuecomment-153423386>
> .
>



-- 
Willian
http://patomandarim.wordpress.com

**Comment by willianfatec** [comment]
I believe the differences are few and quite small. What do you think
@felipemarciel?

2015-11-03 15:23 GMT-02:00 Julius <notifications@github.com>:

> Is it possible to cater for both Brazil and Portugal with one translation,
> or are the dialects too different? I'm generally hoping we can have only
> one version of each language, but it might not be possible.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliusSweetland/OptiKey/issues/148#issuecomment-153424205>
> .
>



-- 
Willian
http://patomandarim.wordpress.com

**Comment by JuliusSweetland** [comment]
**NEW STRINGS** I have added a few new strings to the source (English) resource file. Could someone (on here or Transifex) please provide translations for the following new strings in French, German and Russian (the string is in quote marks, with my explanation straight after - you do not have to translate my explanation):

"Keyboard language: {0}" - *I have split out the language into two settings; this controls the OptiKey keyboard layout and which dictionary to use when typing*
"Keyboard language:" - *same as above*
"User interface language: {0}" - *This is the other setting, which controls which language strings are displayed in the user interface - this is things like the text when you hover over a key, and the language of popup messages. It does not influence the language of the dictionary or the layout of the keys in OptiKey at all*
"User interface language:" - *same as above*
"Simulate physical keystrokes when typing characters:" - *OptiKey attempt to simulate physical presses of the keys on your operating system keyboard layout, i.e. if you have your Windows keyboard layout set to French and you type "e", then OptiKey will figure out which physical presses on your keyboard would be required to output an "e". With some combinations of OptiKey language and OS keyboard layout this could result in difficult to understand output, so this option allows you to turn off the key press simulation and just output plain text. The advantage is that whatever you type in OptiKey will be output "as is", but shortcuts like CTRL+P won't work anymore*

**Comment by JuliusSweetland** [comment]
**Update on new strings** French has been updated. German and Russian still required.

**Comment by ninehundred1** [comment]
I can log into transifix Monday or I can send you translations by mail today, whatever you prefer.

**Comment by JuliusSweetland** [comment]
Post them on here if that's convenient?

**Comment by ninehundred1** [comment]
See below. I translated the last string more along the line of
"Simulate key combinations when typing letters"
As it sounded in German more of you simulating the sound and visuals of a physical keyboard. 
If that is missing the meaning use the second translation which is true to the one you supplied. 


> On Nov 7, 2015, at 4:20 AM, Julius <notifications@github.com> wrote:
> 
> NEW STRINGS I have added a few new strings to the source (English) resource file. Could someone (on here or Transifex) please provide translations for the following new strings in French, German and Russian (the string is in quote marks, with my explanation straight after - you do not have to translate my explanation):
> 
> "Keyboard language: {0}" - I have split out the language into two settings; this controls the OptiKey keyboard layout and which dictionary to use when typing
> "Keyboard language:" - same as above
> 
Tastatursprache:

> "User interface language: {0}" - This is the other setting, which controls which language strings are displayed in the user interface - this is things like the text when you hover over a key, and the language of popup messages. It does not influence the language of the dictionary or the layout of the keys in OptiKey at all
> "User interface language:" - same as above
> 

Menüsprache:

> "Simulate physical keystrokes when typing characters:" - OptiKey attempt to simulate physical presses of the keys on your operating system keyboard layout, i.e. if you have your Windows keyboard layout set to French and you type "e", then OptiKey will figure out which physical presses on your keyboard would be required to output an "e". With some combinations of OptiKey language and OS keyboard layout this could result in difficult to understand output, so this option allows you to turn off the key press simulation and just output plain text. The advantage is that whatever you type in OptiKey will be output "as is", but shortcuts like CTRL+P won't work anymore
> 
1
Simuliere Tastenkombinationen bei Buchstaben-Eingabe 

2
Simuliere physische Tastenanschläge bei Buchstaben-Eingabe

**Comment by JuliusSweetland** [comment]
@ninehundred1 Thank you. I've updated master.

**Comment by davidanton1d** [comment]
Hi! It seems that the link https://www.transifex.com/optikey/ is not working. If it's just a temporary error, please ignore this. I'll check again tomorrow.

Anyhow, I would gladly contribute with Swedish localization if that could be of use. (It might not be a prioritized language, but it's the language i know best :)

I'm a junior programmer in VB.NET, but i would like to contribute with what I could. I'll fork and try out the code.

Really great job you've been doing, got really exited when I first heard of this project. Thought of this (super cheap/open source eye tracking) myself several times!

Cheers //David

**Comment by JuliusSweetland** [comment]
Hi David,

The link is working for me now. I find their website a little hit and miss
sometimes.

It would be great to have you on board. Please make a request to translate
Swedish and I'll approve you.

Glad you like OptiKey!

Regards,
Julius

**Comment by davidanton1d** [comment]
Ok, great. I've made the request now!

2. Here's what a Swedish keyboard looks like:
![image](https://cloud.githubusercontent.com/assets/3294797/11074963/f33b081c-87f2-11e5-897a-53c30d63e06a.png)

3. I'll look around for a word list. I know I've seen a free one somewhere. I'll get back to you on that.

4. Please note the the only diacritics used in everyday words are å/Å, ä/Ä and ö/Ö. These are practically just as common as any other a-z character, so it would be somewhat uncomfortable to have them share keys with other characters. Characters like á, à, é are not very common and does not need their own keys.

**Comment by JuliusSweetland** [comment]
Great info. Thank you. I'll copy this over to a Swedish issue when the
translation is ready to go.

**Comment by thomaskd** [comment]
Hi Julius, great work to you and everyone :) I'm working on the Danish translation.

1. Still working my way through the UI strings, they should be done before too long.

2. Here is a Danish keyboard layout, note that æ,ø, and å are considered letters by themselves and not letters with diacritics, which are not used in Danish words.
![2000px-kb_danish svg](https://cloud.githubusercontent.com/assets/6087348/11093047/1a17edca-8888-11e5-91aa-318a2268d342.png)

3. I'm working on a Danish dictionary, currently around 91k words, I may have to go through some of it manually to remove a lot of the very rarely used words.

**Comment by JuliusSweetland** [comment]
I have added some new settings (and so strings), so could anyone supply French, German, or Russian translations of the following (I will add these back to Transifex soon):

1."Suppress mouse magnification for scrolling actions"
2."Conversation keyboard thickness (left,top,right,bottom in px)"

**Comment by ninehundred1** [comment]
Will do when sober

**Comment by ninehundred1** [comment]
> On Dec 5, 2015, at 3:49 PM, Julius <notifications@github.com> wrote:
> 
> I have added some new settings (and so strings), so could anyone supply French, German, or Russian translations of the following (I will add these back to Transifex soon):
> 
> 1."Suppress mouse magnification for scrolling actions"
> 

1. "Unterdrücke Mausvergrösserung für Scrollfunktionen"

> 2."Conversation keyboard thickness (left,top,right,bottom in px)"
> 

2. "Dicke der Unterhaltungstastatur
(links, oben, rechts, unten in px)"

**Comment by JuliusSweetland** [comment]
@ninehundred1 Great, thank you.

**Comment by JuliusSweetland** [comment]
Could someone please translate these two into French and/or Russian:

1."Suppress mouse magnification for scrolling actions"
2."Conversation keyboard thickness (left,top,right,bottom in px)"

**Comment by JuliusSweetland** [comment]
**ALL** There is a new pre-release of 2.2.2 available here: https://github.com/OptiKey/OptiKey/releases/tag/v2.2.2_Russian_pre-release_3

It includes Russian and I think is nearly read (it is missing the above 2 strings in French and Russian, but is otherwise dev complete). If you could give it a quick test and let me know if there are any major problems that would be hugely appreciated!

**Comment by ninehundred1** [comment]
I will check it out tomorrow.
Also, a friend of mine works for this foundation
http://www.prize4life.org/
do you think it might be good to get you in touch with them or let them
know about what you are making?
Not sure how you handle all the outreach work.

**Comment by JuliusSweetland** [comment]
@ninehundred1 Please do. I've applyied to try and win their grant (details here: http://www.alsa.org/news/media/press-releases/assistive-technology-challenge.html). It's in 2 phases, but imagine how far and fast OptiKey could progress with full-time development. Cross your fingers!

**Comment by ninehundred1** [comment]
I am friends with Neta Zach actually.. So what do you want me to do?

**Comment by JuliusSweetland** [comment]
Ah I emailed her a couple of times to ask questions about the grant. Watched one of her talks on YouTube also. I think she's roughly aware of me and OptiKey, but if you can dance around her with some sort of "OptiKey for the win" banner it couldn't hurt.

**Comment by ninehundred1** [comment]
Ok, I will make sure to mention that what I think you do is great and how
smooth it all works, and maybe the fact that I helped with the translation
also adds to what I think of OptiKey and helps her remember it.

**Comment by JuliusSweetland** [comment]
That'd be amazing. Thanks Stephan!

**Comment by JuliusSweetland** [comment]
@ninehundred1 Actually one of my strings was incorrect. Could you translate this instead: "Conversation keyboard border thickness (left,top,right,bottom in px)"?

**ALL** I realised one of my strings is wrong - could someone please translate these into French and/or Russian:

1."Suppress mouse magnification for scrolling actions"
2."Conversation keyboard border thickness (left,top,right,bottom in px)"

**Comment by ninehundred1** [comment]
"Rahmendicke der Unterhaltungstastatur
(links, oben, rechts, unten in px)"



Assuming frame thickness would also be ok to use.

**Comment by JuliusSweetland** [comment]
Perfect, thanks

**Comment by Nicholas-Mitchell** [comment]
A french translation for the above could be:

1. "Supprimer le groississement lors du défilement"
2. "Épaisseur du cadre du clavier de discussion (gauche, haut, droite, bas, en pixles)"

I say 'could' because there is room for discussion. That will definitely get the point across though.

**Comment by JuliusSweetland** [comment]
Hi Nicholas,

Thank you. I had the following translations supplied. Let me know if they
are ok:

Épaisseur de la bordure du clavier (gauche, haut, droite, bas en pixels)

Suppression de la loupe pendant l'utilisation de l'ascenseur

Regards,
Julius

**Comment by Nicholas-Mitchell** [comment]
Hi Julius,

both are fine, I think. I'd probably take my (1) and their (2). So:
> 1. "Supprimer le groississement lors du défilement"
> 2. "Épaisseur de la bordure du clavier (gauche, haut, droite, bas en
pixels)"

The main difference in (1) is me using "groississement", which really means
'magnification', whereas their "loupe" means 'a magnifying glass'.

The other main difference in (2),  I said "cadre", which means
frame/border, they said "bordure" which is more directly 'border'. Similar
to the German translation I saw using "-Rahmen", which is like 'cadre" in
French. So I guess you could argue for consistency across languages with
that point.

Great work you're doing - not that you need more people telling you that!
:-)

I had a look to see how I might contribute, but my programming skills
aren't really great. I'm more of a statistics guy (in finance)... Maybe at
some point in the future you could do some analysis on usage to try and
optimise further what you have achieved...

Regards,
Nicholas

**Comment by JuliusSweetland** [comment]
Thanks for that Nicholas. I'll take your advice and swap in your first
sentence for the next release.

A bit of statistical analysis sounds like a good plan, but we'd need some
built in telemetry first. It's on my longer to-do list!

**Comment by JuliusSweetland** [comment]
Hi everyone - you may have noticed (or received a notification) that a set of new English strings have been uploaded and need translating. Most languages have dropped from 100% complete to 74% complete. The new strings are for a big new feature which allows users to specify the dwell selection completion time (how long you must look at a key before it selects when using the dwell/fixation trigger) on individual keys. I had to duplicate a lot of existing strings and remove the newline, so most of these new strings are going to need exactly the same translation as before, just without the translation being split over 2 or 3 lines. There are also a few new strings which refer to keyboards, e.g. "NUMERIC & SYMBOL 1", "NUMERIC & SYMBOL 2", etc, plus a couple of new strings to describe the new settings.

The new feature will be included with the next release (2.2.3), which will be released as soon as the French, German and Russian translations are back to 100%.

Thank you for your help.
Julius

**Comment by ninehundred1** [comment]
I'll do that tomorrow

**Comment by JuliusSweetland** [comment]
Great, thanks Stephan.

**Comment by ninehundred1** [comment]
Could you explain what
-destructive keys
does, so I know I have the right translation?
Thanks

**Comment by ninehundred1** [comment]
Also what is the difference between
Mouse (do at point) to Mouse (do at location)?

**Comment by JuliusSweetland** [comment]
Destructive keys = keys which destroy/remove text, e.g. Back one, back
many, clear.

**Comment by JuliusSweetland** [comment]
Do at point means that you select where you want the mouse cursor to move
to and then perform the action, e.g. Left click.

Do at current location means to just perform the action, e.g. Left click,
exactly where the cursor is currently, so the cursor isn't moved before the
action.

**Comment by ninehundred1** [comment]
Ok, done with the german version. Send me a build before you release it so
I can run it through again to see if all sounds fine.

**Comment by JuliusSweetland** [comment]
Excellent, thank you. I will create some pre-releases for sure.

**Comment by Ar2rZ** [comment]
did some(removed spaces from previous) translations, rest of them will translate later once will be able to run app, to understand what you actually added. will try to do that during today, or tomorrow.

**Comment by JuliusSweetland** [comment]
Great, thanks

**Comment by JuliusSweetland** [comment]
New German translations added to master (thank you @ninehundred1). French and Russian to go...

**Comment by Ar2rZ** [comment]
Hi Julius, this string hasn't same meaning - this one:  "Overrides for individual key fixation times (ms):" with this one: "Specify times to complete on individual keys:" as i can't find first one any where..

**Comment by Ar2rZ** [comment]
and i can't find **destructive keys** section..

**Comment by JuliusSweetland** [comment]
In the management console -> Pointing & Selecting this question is asked first: "Specify times to complete on individual keys:" - which means "do you want to specify individual completion times on each key"?
If the check box is ticked, then "Overrides for individual key fixation times (ms):" is the label for the section where the user enters the override completion times, which override the default completion time specified earlier.

Does that clarify things?

**Comment by JuliusSweetland** [comment]
Also the "Destructive Keys" section will be displayed in the same place only if "Specify times to complete on individual keys:" is set to true (checked).

**Comment by Ar2rZ** [comment]
i see now. but destructive keys section still not seen.... 
1. change keyboard keys
2. dock action keys
3. change language keys
4. miscellaneous keys
5. modifier keys
6. Mouse (do at current....)
7. Mouse (do at point)
8.Mouse (misc)
9. move & resize keys
10. Physical keys
11. suggestion keys 

or destructive keys is not a section?

**Comment by JuliusSweetland** [comment]
My mistake - I left in an old group label. I have removed it now and updated Transifex. Thanks.

**Comment by Ar2rZ** [comment]
Just created pull request.

**Comment by JuliusSweetland** [comment]
Just French to go now...

**Comment by JuliusSweetland** [comment]
**ALL** This is a request for any French speakers to complete the outstanding French translations on the Transifex site. German and Russian are complete so after French is brought up to 100% I can release version 2.2.3. Thank you!

**Comment by JuliusSweetland** [comment]
**ALL** There is a pre-release (release candidate) of v2.2.3 available here: https://github.com/OptiKey/OptiKey/releases - if you have a moment could you please give it a test and let me know if you find any problems. The main new feature is the ability to specify selection completion times for individual keys (Management Console -> Pointing and Selecting -> "Specify times to complete on individual keys:" set to true.

**Comment by ninehundred1** [comment]
When I install in German the menu is in English. When I change the language
to German, it is still English in the menu language (including all key
labels). The keyboard is German (umlauts, etc), but the rest is English.
Same when I change to Russian, the Keyboard changes, but the rest stays on
English.

**Comment by JuliusSweetland** [comment]
Hi Stephan,

OptiKey always starts in English the first time it is installed. This is
something that I might change when OptiKey supports more languages.

In the management console you should see two language settings: one
controls the keyboard layout and dictionary being used (it sounds like this
is the setting you are changing). The other setting is the UI language,
which changes the language used on the keys, in pop-up messages and in the
management studio. They can be changed independently so that you can switch
keyboard layouts but keep the UI language consistent. Does that make sense?

**Comment by ninehundred1** [comment]
I forgot that..
Looks good to me, also the functionality now that you can customize those
times.

I suppose there is no need to access/control that management console using
optikey?

**Comment by JuliusSweetland** [comment]
Glad it's all working. I thought some people might want to change OptiKey
settings using OptiKey itself (i.e. People who cannot use a physical mouse
& keyboard) so I made it possible. I don't think it negatively impacts
anything?

**Comment by ninehundred1** [comment]
No, I agree that is useful. I thought it would be even easier to have the window pop up using a key within Optikey such as 'advanced settings' or so, rather than via a keyboard short cut. 
I forgot you could use Optikey to type alt m.

**Comment by JuliusSweetland** [comment]
That, or there's a new right click context menu if you right click on
OptiKey itself. I am still thinking about adding a management console key,
but I'm worried it would be too easy to trigger and make it too easy to
accidentally change settings. I'll keep thinking.

**Comment by JuliusSweetland** [comment]
**Dutch** A pre-release of v2.2.4 is available here: https://github.com/OptiKey/OptiKey/releases

It includes 2 Dutch keyboards for people used to the Belgian keyboard layout or the Netherlands keyboard layout.

If you could test the pre-release and let me know if it looks good (or if you find any issues) then I can promote this to an official release.

Thanks.

**Comment by utacub** [comment]
Hi, we would like to help with the **Spanish** and **Catalan** versions, The keyboard is the same for both languages:
http://cdn.fontmeme.com/images/Spanish-keyboard-550x183.png
If you send us the instructions to translate the UI strings we will gladly do that also.

**Comment by JuliusSweetland** [comment]
@utacub Great stuff - can you switch over to https://github.com/OptiKey/OptiKey/issues/149 and I'll post some details of what is required...

**Comment by morabitina** [comment]
Hi I would like to help with the Spanish version. Please, give me instructions how to procede. Thanks

**Comment by JuliusSweetland** [comment]
@morabitina Hi - thank you for your offer. The Spanish branch is discussed here: https://github.com/OptiKey/OptiKey/issues/149

We are planning to develop Spanish and Catalan separately, and pretty much everything looks to be ready for Spanish. Could you inspect the top post of that issue and let me know (on that issue) if you have any corrections, or anything to add?

**Comment by morabitina** [comment]
Ok! No problem
I am Italian but I live in Spain, so I may can help with Italian too If you
need.
Regards,
Tina

**Comment by Chinesevolunteer** [comment]
Hi，I am Chinese，is chines version be consider？  My English is poor :p

**Comment by JuliusSweetland** [comment]
@Chinesevolunteer Chinese is on the list, but not scheduled yet as there are a lot of complications that need to be considered. I'll let you know when we start on Chinese. Thanks!

**Comment by H1ghty** [comment]
Hi @lukeses is the polish language available already, somewhere?

**Comment by JuliusSweetland** [comment]
@H1ghty Polish is not currently available. I can see that you've requested it on Transifex and I've just approved it. Once the UI strings are complete on Transifex let me know and I'll start a Polish issue to discuss keyboard layout etc. Thanks.

**Comment by JuliusSweetland** [comment]
@H1ghty Actually ignore that. Polish is 100% translated already on Transifex. I will create an issue for it here: https://github.com/OptiKey/OptiKey/issues/238 - please can you give me information about the keyboard layouts etc following the guide at the very top of this page. There are a few languages ahead of Polish at the moment, but we can gather the relevant info and go from there.

**Comment by KrzysztofGoralski** [comment]
Hi @JuliusSweetland , ok. I will try to provide needed informations as soon as possible. :)
But wanted to ask you first what is the 'status' , because @lukeses  was talking about polish language too. Thanks!

**Comment by H1ghty** [comment]
@JuliusSweetland , ok. I will try to provide needed informations as soon as possible. :)
But wanted to ask you first what is the 'status' , because @lukeses was talking about polish language too. Thanks!

**Comment by lukeses** [comment]
@H1ghty I informed about status of the project in #238.

**Comment by JuliusSweetland** [comment]
*ALL* - could you please provide a translation of this new string into French, German, Russian, Dutch and Greek:

*Suppress modifier keys for all mouse actions:*

It is a new setting which sits alongside "Suppress modifier keys when in mouse keyboard:" and means that modifiers (such as SHIFT, ALT, CTRL, WIN) will always be temporarily released when performing a mouse action such as a left click, to prevent the modifier keys changing the behaviour of the click. This is required as I am duplicating the most common mouse actions (left click and left double click) onto the main ALPHA keyboard to speed things up for users.

This is what Google translate comes up with, but these will need correcting - please post whatever translation you know:

French: "Désactive les méta-touches pour toutes les actions de la souris:"
German: "Unterdrückt Zusatztasten für alle Mausaktionen:"
Russian: "Подавить клавиши-модификаторы для всех действий мыши:"
Dutch: "Onderdruk modifier sleutels voor alle muisacties:"
Greek: "Καταστολή πλήκτρα τροποποίησης για όλες τις δράσεις του ποντικιού:"

For reference (and consistency) here are the existing translations for the similar setting "Suppress modified key when in mouse keyboard":
French: "Désactive les méta-touches dans le clavier souris :"
German: "Ignoriere Modifizierungstasten wenn in Maustastatur:"
Russian: "Скрыть модификаторы клавиш если клавиатура мыши активна:"
Dutch: "Onderdruk wijzigingstoetsen wanneer muis toetsenbord wordt gebruikt:"
Greek: "Απόκρυψη πλήκτρων παραμετροποίησης για πληκτρολόγιο ποντικιού"

**Comment by Ar2rZ** [comment]
Hi  Julius,

Have no time to spend on your project at the moment.


Correct translation will be something like this:
Скрыть модификаторы клавиш для всех действий мыши:

Regards,
Arthur

**Comment by andylamp** [comment]
@JuliusSweetland Google translate had me there, better translation given than I came up (due to semantics, but the one given by GT is more to the point. Fixed for both strings (for Greek).

**Comment by JuliusSweetland** [comment]
Is there a PR? I'm at the rugby so I can't check until later... Thanks

**Comment by JuliusSweetland** [comment]
@Ar2rZ Thanks Arthur.

**Comment by JuliusSweetland** [comment]
**ALL** I've just released v2.3.1_pre-release_1 (https://github.com/OptiKey/OptiKey/releases) which isn't quite finished (I'm waiting for a few strings to be translated), but includes new dictionaries for all languages (except English) which have been generated from the 100k (or 200k in the case of Dutch) most common words used in tv shows and films of that language. The layout of the keyboards has also been changed to allow a new way of inputting diacritic marks e.g. accents. Full details on the release notes.

Can you try it out and let me know what you think?

*N.B. If you've already run OptiKey with then you will have to manually delete your customised/user dictionaries for each language (except English as this dictionary has NOT changed) - see here: https://github.com/OptiKey/OptiKey/wiki/Factory-reset-the-dictionary*

**Comment by H1ghty** [comment]
Hi @lukeses How can I start testing polish language?

**Comment by JuliusSweetland** [comment]
**ALL** I still need feedback/corrections on the following strings:

French: "Désactive les méta-touches pour toutes les actions de la souris:"
German: "Unterdrückt Zusatztasten für alle Mausaktionen:"
Dutch: "Onderdruk modifier sleutels voor alle muisacties:"
Greek: "Καταστολή πλήκτρα τροποποίησης για όλες τις δράσεις του ποντικιού:"

**Comment by ninehundred1** [comment]
sorry I missed that, you got it almost correct:

*Suppress modifier keys for all mouse actions:*
German: "Unterdrücke Zusatztasten für alle Mausaktionen:"

**Comment by JuliusSweetland** [comment]
Brilliant, thanks Stephan.

**Comment by JuliusSweetland** [comment]
**ALL** Just French and Dutch to go now:

French: "Désactive les méta-touches pour toutes les actions de la souris:"
Dutch: "Onderdruk modifier sleutels voor alle muisacties:"

**Comment by raponiwalter** [comment]
Thankyou for your job! I'll be available for Italian.
I'm available even for anything web related as a developer.

**Comment by JuliusSweetland** [comment]
Thanks Walter. I'll create a new issue for Italian when we get to it.

**Comment by JuliusSweetland** [comment]
**ALL** Can anyone help @andylamp clean the Greek dictionary? As soon as this last task is done we can release Greek and Turkish, and move on to Spanish.

**Comment by JuliusSweetland** [comment]
ALL - Greek is now out (v2.3.1) and a pre release of Turkish is available on the releases page (as v2.3.2)

**Comment by vdjerek** [comment]
Hi! First of all, thanks for the effort that went to OptiKey, I really appreciate what you did!

I'd like to make a Croatian localization for the OptiKey - I have a team of people willing to help with the translation, while I can work on the rest. I have sufficient experience in programming for the localization as it is outlined in the technical manual. I've just ordered the EyeX, so I'd like to put some serious effort into localization as soon as possible. 

Croatian keyboard is the same as Slovenian, which is in process of being localized. It can be found here: http://ascii-table.com/keyboard.php/234-1

Also, for the dictionary I guess the default dictionary that comes with Android could be used (https://android.googlesource.com/platform/packages/inputmethods/LatinIME/+/master/dictionaries/), which also features the word frequency. 

So, I would kindly ask that you approve a new language (Croatian) to the project on Transifex.

Best,

Vedran

**Comment by vdjerek** [comment]
Ok, I finished translating to Croatian. How does one get it reviewed? Is it necessary for moving on?

**Comment by JuliusSweetland** [comment]
It's not necessary. Can you create a Croatian issue with a summary of the
key info in the first post, like the issues for other languages? I can then
create you a branch to work against

**Comment by vdjerek** [comment]
Great, done! https://github.com/OptiKey/OptiKey/issues/263

**Comment by alexandre-mbm** [comment]
> http://www.winedt.org/Dict/

404 Not Found

> https://invokeit.wordpress.com/frequency-word-lists/

Updated at [FrequencyWords](https://github.com/hermitdave/FrequencyWords/) repository.

Initially he declared that _"[Creative Commons – Attribution / ShareAlike 3.0](http://creativecommons.org/licenses/by-sa/3.0/) license applies to the use of the word lists"_, but currently the repository at Gihub an its [content](https://github.com/hermitdave/FrequencyWords/tree/master/content/2016) is under the [MIT License](https://github.com/hermitdave/FrequencyWords/blob/master/LICENSE).

> I have written a script which can extract the top X from these lists if they are good quality for a specific language, but please note that these lists tend to include a lot of duplicates (with different accents/other diacritics) and have generally been pretty poor quality.

@Julius, there we have a pt-br word-list with 50k. Did you filter from it to your 30k? Can you to share your script with us?

**Comment by OneBadNinja** [comment]
Hey there Julius. An incredible project you have started here!

I would like to contribute by creating a Georgian language localisation. I have already joined at Transifex and have suggested the Georgian language. 
I will happily fill in the UI element translations. I have retrieved a word database containing 310000 words. I can filter these in MySQL to remove single letters. 

I can easily get ahold of the georgian keyboard. It is important to note that the Georgian alphabet has a single case - will this be an issue in the implementation?

I kindly await your approval for Georgian on transifex so that I may begin working
All the best!

**Comment by JuliusSweetland** [comment]
@OneBadNinja Hi, that sounds great. I have approved the language on Transifex. The localisation guide is here: https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md

Have a read and let me know if you have any questions. The main steps are:

1.Complete a Transifex translation
2.Get a wordlist. 310,000 is probably a bit large, but potentially useable. We can test.
3.Identify the keyboard layout and build the keyboards
4.Solve any language specific problems, such as the lack of casing.

**Comment by yhwee0521** [comment]
Hi,
You have started great project. 
Do you have any plan to develop Korean version?
I would like to contribute by anyway. 
If you have any help when you develop Korean version, please let me know to help you. 
All the best!!

**Comment by JuliusSweetland** [comment]
@yhwee0521 I have actually done almost all of the work for Korean already. If you are available to test it and finish any missing translations then we can get it released. If you are technical the branch is here: https://github.com/OptiKey/OptiKey/tree/Korean - it includes the keyboard layouts, symbols are combined and broken apart correctly (I think) and most translations should be there.

The Korean translations are here if you'd like to start with those? https://www.transifex.com/optikey/optikey

**Comment by OneBadNinja** [comment]
Hey Julius, could you please clarify the following sentences?

- CONVERSATION ALPHA 1 & CONVERSATION ALPHA 2 
- Trigger button/key released
- Trigger button/key pressed again

**Comment by JuliusSweetland** [comment]
@OneBadNinja I have just updated the resource on Transifex. The idea of having CONVERSATION ALPHA 1 and 2 has actually been removed. Now there is just CONVERSATION ALPHA. This string is used when specifying a selection time specifically for the key which opens the alpha (letters) keyboard in conversation mode (full screen). It appears only in the management console:

![image](https://user-images.githubusercontent.com/4722561/28239322-ead1ae36-6961-11e7-820b-2e248b51cbe6.png)

"Trigger button/key released" and "Trigger button/key pressed again" - these strings are used in the management console when the selection mode is keyboard key or mouse button to specify how multi-key selections are terminated. The first means that you start a multi-key selection by pressing down a key/button, and end the selection by releasing it. The second string means that you start a multi-key selection by pressing down a key/button, you can then release the key/button, and you end the selection by pressing the key/button again.

![image](https://user-images.githubusercontent.com/4722561/28239334-33e56a90-6962-11e7-9f3e-013eff4b68e9.png)

**Comment by OneBadNinja** [comment]
Thank you Julius, all is clear now!

**Comment by samgled** [comment]
@JuliusSweetland Any chance you could add a male voice to the Japanese localization?

I've no idea how the selection of voices work, if it's automatic based on my computer settings or if it's something that needs to be added within OptiKey but currently, I've only got Haruka. If Ichiro was also an option, that'd be awesome for those wanting their voice to be male.

**Comment by JuliusSweetland** [comment]
Hi Sam, the list of voices is automatically created based on the SAPI
compatible voices that you have installed. Are you able to install Ichiro
through Windows Languages? If you are just restart OptiKey to detect it.

**Comment by samgled** [comment]
Hi Julius,

I wondered if it was something like that. Cheers. I should be able to
figure it out then.

Also, my wife just finished the translations for all of the main interface
keys (she ignored the management console settings translations for the
moment). I'm adding that to my fork and will put in a pull request. The
problem is, I've also added landscape/ portait for the Japanese keyboard
and I'm not sure if you want that yet. If that's the case, just ignore the
pull request, but FYI, the J transifex file has been updated.

Kind regards
Sam Gledhill

**Comment by JuliusSweetland** [comment]
Hi Sam,

Yeah that's great. I'll check it out. Yes I like the idea of a portrait
mode. The idea doesn't lend itself particularly well to most keyboards, but
I can see that Japanese would work - good idea.

No rush, but when do you think the Transifex completion might be complete
so I know roughly what the release date might be.

Cheers,
Julius

**Comment by samgled** [comment]
We're working on the leftovers now. If we don't get it done in the next few hours, it might not happen for a week or so. She really doesn't have much free time at all.

**Comment by JuliusSweetland** [comment]
Ok, no pressure. Thanks to you both!

**Comment by JuliusSweetland** [comment]
FYI Japanese has been completed (thank you @samgled) and released in v2.14.0.

**Comment by Jagoten** [comment]
Hej Julius,

I've noticed few people asked here and helped in polish localisation. I also noticed that on transifex polish language has 100% translation status. Nonetheless polish version is still not supported and the work has stopped somewhere. Could You possibly tell me what's still missing from the polish version to make it happen? I'm not familiar with the whole process but I'm willing to learn and help so I'll gladly take the work from wherever it was stopped and do my best to push it forward.

With regards,
Jacek

**Comment by JuliusSweetland** [comment]
Hi Jacek,

If you can create a new issue for polish then we can discuss everything
there.

First I need to create a new branch, which I have created here:
https://github.com/OptiKey/OptiKey/tree/Polish

You need to fork that branch, clone it into your computer and then follow
the localisation guide to add polish:
https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md

The most important things we will need are:

A) an image of a polish keyboard layout to follow
B) details of any special language rules or accents/diacritic marks/other
symbols
C) a polish dictionary of about 100,000 entries (one per line)

Regards,
Julius

**Comment by tonykor** [comment]
Hello Julius, first of all i'd like to thank you for this great project.

I have tried to implement world prediction using prasage for russian language, i have followed instruction provided here https://github.com/OptiKey/OptiKey/wiki/Next-word-prediction, but seems like prediction not understand cyrillic chars. I had texts saved in UTF-8 before converting to presage db file. I have checked the db file through SQLite DB browser and it seems like working and i see words in cyrillic there, but when typing in the optikey i'm getting chars in incorrect encryption. Does anyone have the same issue? Have anyone teach presage predictor in other chars than latin? Thanks for your cooperation.

**Comment by JuliusSweetland** [comment]
Hi there. Is it possible to debug the predictions coming from presage
before they are passed to the output service? If you break in the get
suggestions method you can move back up the call stack towards presage and
check if the encoding is broken at any point.

**Comment by JuliusSweetland** [comment]
Hi Andrej,

Thank you for your email. I've also received your pull request. Is it
complete? If so I'll aim to test it ASAP. Might not be for a few days as
I'm a bit busy today.

Regards,
Julius

**Comment by siamakmu** [comment]
Hi Julius,
I have completed the translate for Persian (fa) language in transifex.com.
please check and tell me next step. please cooperate for adding this language to OptilKey asap.
--
Best Regards,
Siamak.

**Comment by JuliusSweetland** [comment]
Hi Siamak,

That's great, thank you. This issue describes what is required to add a new
language: https://github.com/OptiKey/OptiKey/issues/148

Recently I have completed work to add Urdu, so hopefully I have already
tackled most of the issues we will have. The Urdu thread is here if you
want to see what was done (screenshots at the bottom of that page show
something close to the final keyboard layouts):
https://github.com/OptiKey/OptiKey/issues/540

Regards,
Julius

**Comment by JuliusSweetland** [comment]
@siamakmu Also if you want to try the (unreleased) Urdu keyboards you can install this version of Optikey: https://github.com/OptiKey/OptiKey/releases/download/v2.24.0_pre-release-10/OptiKeySetup-2.24.0-UrduPakistan.exe

**Comment by siamakmu** [comment]
Thanks for your fast reply.
Actually it has different with Persian language.
Please check these file.

[Persian fa characters.xlsx](https://github.com/OptiKey/OptiKey/files/3263463/Persian.fa.characters.xlsx)

**Comment by JuliusSweetland** [comment]
@siamakmu Thank you for that. Can you explain a bit more about the green/red/white characters? Are you saying all are needed, but green are most common and red are uncommon?

What I need to understand really is which letters should be on the main keyboard, which are less common and can be placed on a 2nd keyboard (equivalent to pressing SHIFT to get to another set of characters), and which diacritic marks are required. I will (later) also need to know which characters the diacritic marks can be used with.

Is the Urdu keyboard layout close to what we are trying to achieve, or is it completely different?

Lastly, I cannot promise that I'll be able to start work on this for the next week or two - I am very busy with Urdu, Optikey in general, and my work.

**Comment by siamakmu** [comment]
Persian language has more letters than English (32 main letters). so usually in Persian keyboards they have to put 6 character in main keys. so they put 5 of them (That I made them green in this excel file) for these keys:
[ = ج
] = چ
\ = پ
; = ک
' = گ
![image](https://user-images.githubusercontent.com/51283486/59067866-ee561880-88c7-11e9-9f2a-712a61cc389c.png)
The default OptiKey keyboard doesn't have these character in main window and just have 26 main English letters.
So one character ( ژ ) left and usually you can type it with Shift + ز ( c )
one of these 6 character can replace with ( ئ ) "m" cause it is not one of our 32 main letters and other 5 characters can combine with + Shift.
So in better word whites and greens (row 1 to 32 in excel file) except red one (row 31 m=ئ) and also row 61 (Shift+c = ژ) are 32 of main FA letters. I mean row 1 to 30 + 32 + 61.

if we need to add 5 main letter to + Shift so we need to remove 6 + Shift in common keyboard. if it is like this so please remove row number 38,40,47,52,59,60. why 6 not 5? if you want to replace row number 31 (red one) with main so you need transfer 5 main character and also row 31 to +shift. 

maybe in 90% Urdu letters are like Persian but it doesn't mean the languages and words are same. also the characters position are completely different.

I think I had done these:
1.Translation of the UI strings
2.We need information about what the keyboard looks like.
3.Alphabet
and just "4.A word list" left. I have to send a list of 75,000 to 100,000 Persian words for you, right?

we have a ALS patient so I need your kindly help and cooperate to make this. you know I translate in transifex.com for just two days!

Any help you need in my side I can do.

**Comment by JuliusSweetland** [comment]
@siamakmu Keyboard layout is going to be the most difficult thing to solve. I think I understand that the letter on the 'm' key is not important as it can be created using the base character+the diacritic (Hamza?). Can you answer which diacritics we will need on the 1st and 2nd keyboards? If you think we need all 32 (or 31 without ئ) then we have space for 1 (or 2) diacritics on the first keyboard. If we should have more diacritics then can any of the characters on the 1st keyboard be moved to the second keyboard?

**Comment by siamakmu** [comment]
Sorry Julius, I thought the keyboard must made as English with 26 letter. I checked the Russian Keyboard and it's number of letters is almost near to Persian. now I understand we can have all 32 main letters can be on the main keyboard layout.
I will make and image from Optikey and put all characters on its true position to better understand.

**Comment by JuliusSweetland** [comment]
Perfect thank you. Please note that we CAN put all keys, including
diacritic marks, on one keyboard, but it makes the size of the keys
smaller. If some of the characters are used less frequently we could put
them on a second keyboard like I have done with Urdu. This means we would
have fewer keys on each keyboard and so the keys would be slightly larger.
Let me know what you think.

**Comment by JuliusSweetland** [comment]
@siamakmu Please continue discussion on this issue: https://github.com/OptiKey/OptiKey/issues/568

**Comment by rolivella** [comment]
Hi Julius, 

Thank you for this very nice project, my father is using it and is working fine. 

I work as bioinformatician and I would like to finish the Catalan and Spanish translations. I've checked Transifex (new to me) and Catalan is at 83.8% and Spanish at 69%. Could you guide me on how to complete them? Thank you.

Roger

**Comment by JuliusSweetland** [comment]
@rolivella Hi Roger. That's great, thanks for letting me know. Yes, any help to complete Catalan and Spanish would be very much appreciated. Have a look at the location guide here: https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md

The section you want is at the top titled "UI Translation" and describes how to sign up to Transifex.com and translate the remaining strings.

If you see any incorrect translations you can also correct them on the site. Once you are finished please let me know and I will grab your translations and release a new version of OptiKey with the improved translations. Thanks!

**Comment by rolivella** [comment]
@JuliusSweetland great! I'll work on it and let you know ASAP when finished.

**Comment by JuliusSweetland** [comment]
Thanks Roger

**Comment by lsaranto** [comment]
I would like to translate the finnish language. I've requested the language and it's been accepted. When I go to join the team Finnish is not on the list.

**Comment by JuliusSweetland** [comment]
> 
> 
> I would like to translate the finnish language. I've requested the language and it's been accepted. When I go to join the team Finnish is not on the list.

@Isaranto Hi - I have accepted this request again. Please try now and email me if you have any questions, or when you are finished translating and we can discuss the next steps, i.e keyboard layouts, word lists, etc.

**Comment by JuliusSweetland** [comment]
@Isaranto Hi, I can see the Finnish translation on Transifex is complete. Please let me know if you are available to continue working to create the keyboard(s) etc?

**Comment by isaranto** [comment]
Hi! You have the wrong email address! I think you are looking for LSARANTO
while my starts with an i (isaranto).

**Comment by JuliusSweetland** [comment]
@isaranto Ah apologies. Thanks for letting me know!
@lsaranto Hi, I can see the Finnish translation on Transifex is complete. Please let me know if you are available to continue working to create the keyboard(s) etc?

**Comment by lsaranto** [comment]
@JuliusSweetland Yes, I just finished fixing some mistakes I realised I had made. The translation is now finished and ready. I can try to do the rest of the steps, too.

**Comment by JuliusSweetland** [comment]
Sounds great. Please let me know if you have any questions.

**Comment by lsaranto** [comment]
@JuliusSweetland First a question about the needed word list. Is this ok [http://kaino.kotus.fi/sanat/nykysuomi/](http://kaino.kotus.fi/sanat/nykysuomi/) ? I don't know anything about licenses.

> THE WORD LIST OF THE INSTITUTE FOR THE LANGUAGES OF FINLAND
The Institute for The Languages of Finland publishes a glossary of contemporary Finnish words supplemented with spelling information. The word list is not an exhaustive or authoritative list of Finnish words, but it is intended for example to act as an aid in the development of Finnish language computer programs and Finnish language user interfaces.
The list is published under the GNU LGPL (Lesser General Public License), EUPL v.1.1 (European Union General License) and CC Attribute 3.0 Unmodified. The scope of the word list is 94,110 words. The word list is in XML format and the character set is UTF-8. If the list is later changed, the version number of the list changes.

Then the problems. I was following the guide to add the .resx and .dic files to the VS project. I can't find any of the things mentioned in the steps. I'm using VS2019 Community.

**Comment by JuliusSweetland** [comment]
@lsaranto That word list looks great, but we'd need to extract just the words - one per line. Reading the details a bit it sounds like there is some meta information about each entry: can you check if all the words are complete on their own, or if this list gives enough information so that a user can recreate a much larger word list from the stems + endings (for example)?

And yes, the licenses are compatible.

Regarding your issues: the resx files are the language specific Resource.xx-XX.resx files in the Properties folder (https://github.com/OptiKey/OptiKey/tree/master/src/JuliusSweetland.OptiKey/Properties)

The dic files are the language specific word lists found in the Dictionaries folder (https://github.com/OptiKey/OptiKey/tree/master/src/JuliusSweetland.OptiKey/Dictionaries)

**Comment by lsaranto** [comment]
@JuliusSweetland Yes, I have extracted just the words from the list. The words in the list are in their basic form. I don't know if you are familiar with the Finnish language, but the Finnish language doesn't use for example prepositions. Instead, endings are added to the basic form of the word. So decoding the meta information to each word would make the list extremely large. Also, the given key for the meta information doesn't look like it's complete. Anyway, it's beyond my skills and knowledge.

I'm talking about the steps in [this guide](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md). The guide talks about creating a new branch. Do I need to do it and how?

When trying to do the following:

> After placing the file in the correct folder go back in Visual Studio in OptiKey solution and do the following:
> 
> Right-click JuliusSweetland.OptiKey project
Select Add
Select Existing Item
Go to src\JuliusSweetland.OptiKey\Properties and select the Resources.el-GR.resx file you just placed
Click OK to add the resource file
The final step is to place the Resources.el-GR.resx file inside the Properties folder in the project (which has a wrench icon), do that by dragging and dropping the file inside the solution to the correct place.
> 
> Please ensure that the property of the file named Build Action is set to Embedded Resource.

 I can't find any of these.

**Comment by JuliusSweetland** [comment]
@lsaranto Ok I have created the new branch (you wouldn't have been able to do this on my repository): https://github.com/OptiKey/OptiKey/tree/Finnish

Would it be easier to email me the files and I'll add them to the solution and then check it back into the new Finnish branch. After that I'll let you know and you can clone my branch to complete the other steps? You can email them to me at julius@optikey.org

**Comment by lsaranto** [comment]
I've made the changes and now trying to push. For some reason I get an error:
> Pushing Finnish
Remote: Permission to OptiKey/OptiKey.git denied to lsaranto.
Error encountered while pushing to the remote repository: Git failed with a fatal error.
unable to access 'https://github.com/OptiKey/OptiKey.git/': The requested URL returned error: 403
Pushing to https://github.com/OptiKey/OptiKey.git

Sorry for being such a noob.

**Comment by JuliusSweetland** [comment]
No problem at all. So that error like like you've cloned MY repo (the first
OptiKey in that URL is my team name, the second is the repo name), made
changes, and more are trying to push your changes back to MY repo. That's
not allowed as I own that repo. Instead you need to:

A) physically copy the files on your computer to a different location so
you won't lose your changes
B) FORK my project. It will create a new project, also called OptiKey, but
under your GitHub account. It will be called something like
lsaranto/OptiKey.git
C) Clone your repo down onto your computer
D) Change the files again. You can do this by copying the files from step A
into the new repo location on your computer. Don't copy the .git folder
over though.
E) you should hopefully be have a local copy of your own repo with the
changes you made to mine.
F) Push this to your repo on GitHub.
G) Create a pull request back to my repo.

All very complex so don't worry if you get stuck, just tell me where you
are and I'll try to help!

**Comment by JuliusSweetland** [comment]
@lsaranto Please could you test the pre-release of v2.25.0 to see if Finnish works as expected: https://github.com/OptiKey/OptiKey/releases

**Comment by lsaranto** [comment]
@JuliusSweetland Yes, it's ok.

**Comment by JuliusSweetland** [comment]
@lsaranto Thanks. I'll release it asap.

**Comment by sergioschina** [comment]
Hi @JuliusSweetland , My Uncle has ALS, and I have successfully installed the optikey for him, and I am a good fan of your work and all community on this project. 

Our main language is Brazilian Portuguese, I see, today Optikey only support Portuguese from Portugal. My uncle has asked me if I could change de voice speech to a male voice. So, I saw on Windows 10 Microsoft support website, that only Brazilian Portuguese has support for male voice and the name is Daniel.

![Capturar](https://user-images.githubusercontent.com/55419281/73809480-28a6eb00-47b2-11ea-84e4-ba15aa315cb3.PNG)


How can I change or add a Brazilian Portuguese voice on optikey?

**Comment by JuliusSweetland** [comment]
@sergioschina Hi. I've had a look and it doesn't look like Microsoft have made a male Portuguese (Portugal) voice available. However, there is a voice using eSpeak, which Optikey supports. I can't vouch for its quality, but it is listed here: http://espeak.sourceforge.net/languages.html

Check out the "eSpeak and MBROLA" section of this link for details on how to install and use eSpeak voices: https://github.com/Optikey/Optikey/wiki/Speech

**Comment by kevinlin18** [comment]
Hi! Is that possible to separate 'Language Enum' current used by UI language and keyboard language? or a PR about it is welcome?
There are 3 English options: Canada, UK, US actually shared same UI, the difference is dictionary. I think we cam merge them in UI list, keep separate in keyboard list.

I translated UI into Chinese Traditional and tried to make bopomofo IME work in Optikey.
Localisation is 100% worked, but IME is not in ideal condition.
So I think we can make a PR about UI language first.
But it will crash if someone choose Chinese on keyboard list, so this the second reason I think we should have a new Enum.

**Comment by Chinesevolunteer** [comment]
您好，已收到您的来件。
